add kuksa-cpp-client

AkhilTThomas opened the pull request:
A C++ based client interfacing with kuksa data broker using V1 and V2 API is added.
@SebastianSchildt @nayakned

AkhilTThomas committed:
add kuksa-cpp-client

lint

lukasmittag commented on kuksa-cpp-client/data/vss-core/vss_release_2.0.json:
is it possible to reuse the files from kuksa-common? https://github.com/eclipse-kuksa/kuksa-common

AkhilTThomas replied:
Wouldn't it make more sense if kuksa-common is a submodule of the incubation? that way all the projects can access it?

lukasmittag replied:
Yes, could be a good solution.

SebastianSchildt replied:
I mergend the PR #23 , so I think the "data/vss" files can likely be removed form this PR when it is rebased

SebastianSchildt commented:
Hi @AkhilTThomas , cool this is landing here. I warned you about Linters :D 

Pre-commit is unhappy. See here  https://github.com/eclipse-kuksa/kuksa-incubation?tab=readme-ov-file#pre-commit-set-up how to set up and see the thread/comment here https://github.com/eclipse-kuksa/kuksa-incubation/pull/18#issuecomment-2606888835 how to repair

lukasmittag commented on kuksa-cpp-client/proto/kuksa/val/v1/types.proto:
same for the proto files is it suitable to have them pulled as submodule? Then you do not need to keep track if they change on the master branch :)

AkhilTThomas replied:
@lukasmittag @SebastianSchildt 
I can see that the proto files are also duplicated in multiple incubation projects, considering that proto is a fundamental part of the whole `kuksa` world it can also be part of the common repo. What do you think?

That would also mean making this change in the `kuksa-databroker` repo and keeping the `kuksa-common` as the single source of truth.

SebastianSchildt replied:
Also pinging @rafaeling  and @erikbosch : I do remember we discussed i th past, that something like this is "the right way" to do it evenutally, but Idid not fully recollect what the outcome was. Is the time maybe now?

erikbosch replied:
Once upon the there were some people that had very strong opinion on that the `*.proto` files were "owned" by databroker, i.e. they should not be part of a "common" repo. Those persons are no longer active in Kuksa, so I think it would be great moving them to https://github.com/eclipse-kuksa/kuksa-common

Then, whether you should use a git submodule or some other mechanism like manual clone/checkout as part of setup I do not really have any strong opinion.

SebastianSchildt replied:
Protos exist in https://github.com/eclipse-kuksa/kuksa-proto now and I gave a "v1" tag, because we need to start somewhere. @AkhilTThomas  Maybe easiest for now is adding this via another submodule (we have not yet though about whether we should do downloadabel "releases" and how they should look like....)

AkhilTThomas commented:
> Hi @AkhilTThomas , cool this is landing here. I warned you about Linters :D
> 
> Pre-commit is unhappy. See here https://github.com/eclipse-kuksa/kuksa-incubation?tab=readme-ov-file#pre-commit-set-up how to set up and see the thread/comment here [#18 (comment)](https://github.com/eclipse-kuksa/kuksa-incubation/pull/18#issuecomment-2606888835) how to repair

Yes! Surprisingly the pre-commit is not complaining locally. Might be conflicting with the ones i had in my repo.

AkhilTThomas committed:
fix submodule not showing up

lukasmittag commented on kuksa-cpp-client/ReadMe.md:
```diff
@@ -0,0 +1,51 @@
+    Establish connection to a databroker end-point
+- `get("Vehicle.Speed")`
+    Get a datapoint value
+- `set("Vehicle.Speed",value)`
```
this should be called publish due to the terminology of kuksa.

AkhilTThomas replied:
I have modified this and added better documentation

lukasmittag reviewed (changes requested):
Some documentation about what from v2 is supported by this cpp client would be nice. For example the bi directional streaming part is not supported.

AkhilTThomas committed:
Improve doc and add more APIs

- Added doc strings for v2 APIs
- Support getServerInfo
- Support getValues
- Extend v2 examples

lukasmittag reviewed (approved):
LGTM

SebastianSchildt commented:
Merging this. Hoping for more updates in the future :D
